File: argon-io/jvm/src/main/scala/dev/argon/io/PathBinaryResource.scala
package dev.argon.io

import dev.argon.io.*
import dev.argon.nobleidl.runtime.util.{InputStreamWithError, RefiningInputStreamWithError}
import dev.argon.util.async.ErrorWrapper
import zio.*
import zio.stream.ZStream

import java.io.{IOException, InputStream}
import java.nio.channels.SeekableByteChannel
import java.nio.file.{Files, Path}


private[io] final class PathBinaryResource(path: Path) extends BinaryResource[IOException] {
  override def asBytes: ZStream[Any, IOException, Byte] =
    ZStream.fromPath(path).refineToOrDie[IOException]

  override def asInputStream[E1 >: IOException](using ew: ErrorWrapper[E1] { type EX <: IOException }): ZIO[Scope, ew.EX, InputStreamWithError[ew.EX]] =
    ErrorWrapper.wrapEffect[Scope, E1, InputStream](
      ZIO.fromAutoCloseable(ZIO.attempt { Files.newInputStream(path).nn }
        .refineToOrDie[IOException])
    )
      .map { is =>
        new RefiningInputStreamWithError[ew.EX](is) {
          override def wrapIOException(ex: IOException): RuntimeException =
            throw ew.wrap(Cause.fail(ex))
        }
      }

  override def fileName: Option[String] = Some(path.toString)
}

File: launcher/launcher/jvm/src/main/java/dev/argon/driver/launcher/PathBinaryResource.java
package dev.argon.driver.launcher;

import dev.argon.backend.api.BinaryResource;
import dev.argon.nobleidl.runtime.util.IOErrorType;
import dev.argon.nobleidl.runtime.util.InputStreamWithError;

import java.io.IOException;
import java.io.InterruptedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;
import java.util.function.Function;

class PathBinaryResource extends BinaryResource<IOException> {
	public PathBinaryResource(Path path) {
		this.path = path;
	}

	private final Path path;

	@Override
	public Optional<String> fileName() {
		return Optional.ofNullable(path.getFileName()).map(Path::toString);
	}
	
	@Override
	public <EE extends IOException> InputStreamWithError<EE> asInputStream(IOErrorType<IOException, EE> errorType) throws InterruptedIOException, EE {
		try {
			try {
				var is = Files.newInputStream(path);
				return InputStreamWithError.ofInputStream(is)
					.convertError(
						IOErrorType.fromClass(IOException.class),
						Function.identity(),
						errorType
					);
			}
			catch(IOException e) {
				throw errorType.toThrowable(e);
			}
		}
		catch(InterruptedException e) {
			var ex = new InterruptedIOException(e.getMessage());
			ex.setStackTrace(e.getStackTrace());
			throw ex;
		}
	}
}
